File: core/src/test/scala/diplomacy/mock/rule/MockRule.scala
package diplomacy.mock.rule

import diplomacy.rule.{ InvalidOrderMessage }
import diplomacy.mock.board.{ MockPower, MockMilitaryBranch }

abstract class MockRule extends Rule {

  protected[this] def resolveProcedure(
    board: Board, orders: Set[Order]
  ): Either[InvalidOrderMessage, ResolvedResult] = ???

  protected[this] def unitsRequiringOrder(board: Board): Set[DiplomacyUnit] = Set()

  protected[this] def errorMessageOfOrder(board: Board)(
    order: Order
  ): Option[InvalidOrderMessage] = None
  protected[this] def errorMessageOfOrders(board: Board)(
    order: Set[Order]
  ): Option[InvalidOrderMessage] = None

  protected[this] def defaultOrderOf(board: Board)(unit: DiplomacyUnit): Order = ???
}

File: test/mock/mock-rule.ts
import { rule } from "../../src/rule"
import { board } from "../../src/board"
import { util } from "../../src/util"
const { Failure } = util

export class MockRule extends rule.Rule<string, board.Name, string, string, string, string, string> {
  protected unitsRequiringOrder (
    board: board.Board<string, board.Name, string, string, string>
  ) {
    return new Set()
  }
  protected defaultOrderOf (
    board: board.Board<string, board.Name, string, string, string>,
    unit: board.Unit<string, board.Name>
  ) {
    return null
  }
  protected errorOfOrder (
    board: board.Board<string, board.Name, string, string, string>,
    order: rule.Order<string, board.Name>
  ) {
    return null
  }
  protected errorOfOrders (
    board: board.Board<string, board.Name, string, string, string>,
    orders: Set<rule.Order<string, board.Name>>
  ) {
    return null
  }
  protected resolveProcedure (
    board: board.Board<string, board.Name, string, string, string>,
    orders: Set<rule.Order<string, board.Name>>
  ) {
    return new Failure("")
  }
}
